### PR TITLE
Add asset entities to graphql route #4521

### PR DESCRIPTION
**Pull request by agzam**
> **Epic** #
> Close 
> Related https://github.com/threatgrid/iroh/issues/4521

*Note: this is a multi-PR work. The second part is [here #1068 ](https://github.com/threatgrid/ctia/pull/1068)*

<a name="qa">[§](#qa)</a> QA
============================

1. Create one or multiple Assets by issuing requests to `POST ​/ctia​/asset` route
2. Try running simple graphql queries, e.g.: 

```
POST  /ctia/graphql
{
"query": "query Assets { assets { nodes {id} }}"
}
```

3. Make sure the results include all (previously created) assets
4. Similarly, test it for `AssetMapping`, `AssetProperties` and `TargetRecord`

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: graphql queries for Asset, AssetMapping, AssetProperties and TargetRecord entities
public: one line description for public release notes.
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

2d651cc4 adds an asset-mapping specific test case
95357042 remove unused code
43710f3f adds target-record graphql tests
b97bc4f7 adds target-record graphql schema
6581d90c remove singular form of asset-properties
2b084878 adds AssetProperties tests
0ef440bf normalize ns headers
3785a4e9 adds asset-properties graphql schema
609fe710 adds basic asset-mapping graphql test
16bc4277 Adds graphql schemas for asset-mapping
214f34c1 extend asset graphql tests
883b3e3f move asset graphql schemas to its own namespace
657f7363 adds additional graphql schema things
7254460c adds AssetType to graphql/schemas
dd0f4d03 normalize and sort ns headers

### CODE REVIEW

**Comment by ereteog** [comment]
quick tip, for your failing request request, you can exercise it with the `graphiql` interface: `http://localhost:3000/graphiql`. In particular, you have the graphql model displayed in the right panel, starting from the root node.

**Comment by agzam** [comment]
> quick tip, for your failing request request, you can exercise it with the `graphiql` interface: `http://localhost:3000/graphiql`. In particular, you have the graphql model displayed in the right panel, starting from the root node.

This is a good tip. Thank you @ereteog

**Review by ereteog — COMMENTED** [review]
It seems ok. You can similarly add asset-mapping, assert-properties and target-record.

**Comment by agzam** [comment]
@ereteog This one is almost baked.  There are couple of things I need to address (perhaps I can do it in a separate PR)

- first, is the relationship stuff you mentioned (I'm not sure yet how to set that right)
- second, I did a sloppy job replicating things based on existing code. asset-properties do not have a singular form, so, some parts of the code are disposable (query and relevant bits of asset-property-test), I need to figure out how to straighten that up

As always - any pointers, comments, suggestions are welcome.

**Comment by ereteog** [comment]
@agzam this is part of this task to ad

> @ereteog This one is almost baked. There are couple of things I need to address (perhaps I can do it in a separate PR)
> 
>     * first is the relationship stuff you mentioned (I'm not sure yet how to set that right)

This is part of this task but you can proceed as it suits you :-). If you open a second PR do not close the task with this one.

>     * second, I did a sloppy job replicating things based on the existing code. `asset-properties` do not have a singular form, so, some parts of the code are disposable (query and relevant bits of asset-property-test), I need to figure out how to straighten that up

I am not sure what is problematic in your PR for this, could you add some comment in the PR?

> As always - any pointers, comments, suggestions are welcome.

for the connections between `asset` and its related entity types, you can simply replicate the mechanism for `feedback` and its `entity_id` field (see how `feedback/feedback-connection-field` works). The goal will be to retrieve from the graphql endpoint assets and corresponding mappings and properties with a single query.

**Comment by agzam** [comment]
@ereteog I've attempted to gain some understanding how Feedback and entities get linked and tried to replicate it for `Asset's` relationship to `AssetMapping` and `AssetProperties` via `asset_ref`, see: https://github.com/threatgrid/ctia/pull/1064/commits/e7aaa5bdd3267dcaf6a16d8628f1fe2ee8a12e83 
I think this can get a bit hairy, I should probably do it in a separate PR. Would you approve this PR (I will remove aforementioned commit) without the `asset-ref-connection-field`? I'll resume my efforts in a separate PR.

**Comment by ereteog** [comment]
@agzam like I said:

> This is part of this task but you can proceed as it suits you :-). If you open a second PR do not close the task with this one.

So, if you prefer to proceed in 2 PRs, add the missing entity `target-record`, complete the QA section for other entities, and replace `close`by `related` to not close the issue.

File: src/ctia/graphql/schemas.clj [hunk 1]
```diff
@@ -135,6 +136,14 @@
                                        sighting/sighting-order-arg
                                        p/connection-arguments)
                        :resolve (res/search-entity-resolver :sighting)}
+    :target_record    {:type    TargetRecordType
+                       :args    search-by-id-args
+                       :resolve (res/entity-by-id-resolver :target-record)}
```
Comment (agzam) [inline src/ctia/graphql/schemas.clj:141]: @ereteog I'm stuck. I've been trying to figure it out, but I'm really stuck.

For some reason the last commit is breaking things and I'm not sure exactly what I'm missing (TargetRecord should almost be like AssetMapping, but as soon I uncomment the highlighted lines, I can't start the server anymore, it fails with:

    2021-02-01 18:24:43,380 ERROR (async-dispatch-2) [puppetlabs.trapperkeeper.internal] - Error during service start!!! graphql.AssertException: type Verdict not found in schema at graphql.Assert.assertTrue(Assert.java:41) at graphql.schema.SchemaUtil.resolveTypeReference(SchemaUtil.java:244) at graphql.schema.GraphQLList.replaceTypeReferences(GraphQLList.java:45) at graphql.schema.SchemaUtil.resolveTypeReference(SchemaUtil.java:248) at graphql.schema.GraphQLFieldDefinition.replaceTypeReferences(GraphQLFieldDefinition.java:69) at graphql.schema.SchemaUtil.resolveTypeReferencesForFieldsContainer(SchemaUtil.java:228) at graphql.schema.SchemaUtil.replaceTypeReferences(SchemaUtil.java:212) at graphql.schema.GraphQLSchema$Builder.build(GraphQLSchema.java:300) at graphql.schema.GraphQLSchema$Builder.build(GraphQLSchema.java:289) at ctia.schemas.graphql.helpers$eval52786$new_schema__52791$fn__52792$fn52793__52803$fn__52804.invoke(helpers.clj:488)
        at ctia.schemas.graphql.helpers$eval52786$new_schema__52791$fn__52792$fn52793__52803.invoke(helpers.clj:486)
        at clojure.lang.AFn.applyToHelper(AFn.java:154)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at ctia.schemas.core$eval34569$resolve_with_rt_ctx__34574$fn__34575.invoke(core.clj:158)
        at ctia.schemas.core$eval34569$resolve_with_rt_ctx__34574.invoke(core.clj:152)
        at ctia.schemas.graphql.helpers$eval52831$new_graphql__52836$fn__52837$fn52838__52848$fn__52849.invoke(helpers.clj:502)
        at ctia.schemas.graphql.helpers$eval52831$new_graphql__52836$fn__52837$fn52838__52848.invoke(helpers.clj:500)
        at clojure.lang.AFn.applyToHelper(AFn.java:154)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:31)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at ctia.schemas.core$eval34569$resolve_with_rt_ctx__34574$fn__34575.invoke(core.clj:158)
        at ctia.schemas.core$eval34569$resolve_with_rt_ctx__34574.invoke(core.clj:152)
        at ctia.http.server_service_core$eval86545$start__86550$fn__86551.invoke(server_service_core.clj:40)
        at ctia.http.server_service_core$eval86545$start__86550.invoke(server_service_core.clj:24)
        at ctia.http.server_service$reify__86646$service_fnk__12275__auto___positional$reify__86658.start(server_service.clj:21)
        at puppetlabs.trapperkeeper.services$eval12073$fn__12087$G__12063__12090.invoke(services.clj:9)
        at puppetlabs.trapperkeeper.services$eval12073$fn__12087$G__12062__12094.invoke(services.clj:9)
        at puppetlabs.trapperkeeper.internal$eval22325$run_lifecycle_fn_BANG___22332$fn__22333.invoke(internal.clj:196)
        at puppetlabs.trapperkeeper.internal$eval22325$run_lifecycle_fn_BANG___22332.invoke(internal.clj:179)
        at puppetlabs.trapperkeeper.internal$eval22354$run_lifecycle_fns__22359$fn__22360.invoke(internal.clj:229)
        at puppetlabs.trapperkeeper.internal$eval22354$run_lifecycle_fns__22359.invoke(internal.clj:206)
        at puppetlabs.trapperkeeper.internal$eval22982$build_app_STAR___22991$fn$reify__23003.start(internal.clj:605)
        at puppetlabs.trapperkeeper.internal$eval23030$boot_services_for_app_STAR__STAR___23037$fn__23038$fn__23040.invoke(internal.clj:631)
        at puppetlabs.trapperkeeper.internal$eval23030$boot_services_for_app_STAR__STAR___23037$fn__23038.invoke(internal.clj:629)
        at puppetlabs.trapperkeeper.internal$eval23030$boot_services_for_app_STAR__STAR___23037.invoke(internal.clj:623)
        at clojure.core$partial$fn__5841.invoke(core.clj:2630)
        at puppetlabs.trapperkeeper.internal$eval22399$initialize_lifecycle_worker__22410$fn__22411$fn__22562$state_machine__19592__auto____22587$fn__22590.invoke(internal.clj:249)
        at puppetlabs.trapperkeeper.internal$eval22399$initialize_lifecycle_worker__22410$fn__22411$fn__22562$state_machine__19592__auto____22587.invoke(internal.clj:249)
        at clojure.core.async.impl.ioc_macros$run_state_machine.invokeStatic(ioc_macros.clj:978)
        at clojure.core.async.impl.ioc_macros$run_state_machine.invoke(ioc_macros.clj:977)
        at clojure.core.async.impl.ioc_macros$run_state_machine_wrapped.invokeStatic(ioc_macros.clj:982)
        at clojure.core.async.impl.ioc_macros$run_state_machine_wrapped.invoke(ioc_macros.clj:980)
        at clojure.core.async$ioc_alts_BANG_$fn__19839.invoke(async.clj:421)
        at clojure.core.async$do_alts$fn__19770$fn__19773.invoke(async.clj:288)
        at clojure.core.async.impl.channels.ManyToManyChannel$fn__14327$fn__14328.invoke(channels.clj:95)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at clojure.core.async.impl.concurrent$counted_thread_factory$reify__14196$fn__14197.invoke(concurrent.clj:29)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.base/java.lang.Thread.run(Thread.java:832)

**Review by ereteog — APPROVED** [review]
LGTM.

I would strengthen the QA section with for more specific examples and more fields.
For instance for target record you could propose to retrieve `targets.observables.type`.

**Comment by hisans** [comment]
QA Verified in 1.66 Release

**Pull request merged**